fs/fs_zenfs.cc

```diff
@@ -794,6 +794,81 @@ IOStatus ZenFS::GetFileSize(const std::string& f, const IOOptions& options,
   return s;
 }
 
+/* Must hold files_mtx_ */
+IOStatus ZenFS::RollbackAuxDirRenameNoLock(
+    const std::string& source_path, const std::string& dest_path,
+    const std::vector<std::string>& renamed_children, const IOOptions& options,
+    IODebugContext* dbg) {
+  IOStatus s;
+
+  for (const auto& rollback_child : renamed_children) {
+    std::string source_rollback_child_path =
+        (std::filesystem::path(dest_path) /
+         std::filesystem::path(rollback_child))
+            .string();
+    std::string dest_rollback_child_path =
+        (std::filesystem::path(source_path) /
+         std::filesystem::path(rollback_child))
+            .string();
+    s = RenameFileNoLock(source_rollback_child_path, dest_rollback_child_path,
+                         options, dbg);
+    if (!s.ok()) {
+      return IOStatus::Corruption(
+          "RollbackAuxDirRenameNoLock: Failed to roll back directory rename");
+    }
+  }
+
+  s = target()->RenameFile(ToAuxPath(dest_path), ToAuxPath(source_path),
+                           options, dbg);
+  if (!s.ok()) {
+    return IOStatus::Corruption(
+        "RollbackAuxDirRenameNoLock: Failed to roll back auxiliary path "
+        "renaming");
+  }
+
+  return s;
+}
+
+/* Must hold files_mtx_ */
+IOStatus ZenFS::RenameAuxPathNoLock(const std::string& source_path,
+                                    const std::string& dest_path,
+                                    const IOOptions& options,
+                                    IODebugContext* dbg) {
+  IOStatus s;
+  std::vector<std::string> children;
+  std::vector<std::string> renamed_children;
+
+  s = target()->RenameFile(ToAuxPath(source_path), ToAuxPath(dest_path),
+                           options, dbg);
+  if (!s.ok()) {
+    return s;
+  }
+
+  GetZenFSChildrenNoLock(source_path, true, &children);
+
+  for (const auto& child : children) {
+    std::string source_child_path =
+        (std::filesystem::path(source_path) / std::filesystem::path(child))
+            .string();
+    std::string dest_child_path =
+        (std::filesystem::path(dest_path) / std::filesystem::path(child))
+            .string();
+    s = RenameFileNoLock(source_child_path, dest_child_path, options, dbg);
```

Comment:
This code is repeated in RollbackAuxDirRenameNoLock() and here. One suggestion is can we simplify it by moving it to a separate function ?
pseudo code:
```
IOStatus UpdatePathAndRename(std::string src, std::string dest, std::vector<std::string> filelist, IOOptions& options, IODebugContext& dbg ) {
IOStatus s;
  for (const auto& child : filelist) {
    std::string source_child_path =
        (std::filesystem::path(src) / std::filesystem::path(child))
            .string();
    std::string dest_child_path =
        (std::filesystem::path(dest) / std::filesystem::path(child))
            .string();
    s = RenameFileNoLock(source_child_path, dest_child_path, options, dbg);
    if (!s.ok()) return s;
  }
  return s;
}

```

Remaining code can be same...

Reply:
This does not seem to simplify the code because of  `  renamed_children.push_back(child);` inside this for loop.

Reply:
I think @aravind-wdc suggests something like this:
```c++
IOStatus ZenFS::RenameChildren(std::string const& src_dir, std::string const& dest_dir,
            std::string const& child, const IOOptions& options,
            IODebugContext* dbg) {
  std::string source_rollback_child_path =
      (std::filesystem::path(src_dir) / std::filesystem::path(child)).string();
  std::string dest_rollback_child_path =
      (std::filesystem::path(dest_dir) / std::filesystem::path(child)).string();
  return RenameFileNoLock(source_rollback_child_path, dest_rollback_child_path,
                       options, dbg);
}

/* Must hold files_mtx_ */
IOStatus ZenFS::RollbackAuxDirRenameNoLock(
    const std::string& source_path, const std::string& dest_path,
    const std::vector<std::string>& renamed_children, const IOOptions& options,
    IODebugContext* dbg) {
  IOStatus s;

  for (const auto& rollback_child : renamed_children) {
    s = RenameChildren(dest_path, source_path, rollback_child, options, dbg);
    if (!s.ok()) {
      return IOStatus::Corruption(
          "RollbackAuxDirRenameNoLock: Failed to roll back directory rename");
    }
  }

  s = target()->RenameFile(ToAuxPath(dest_path), ToAuxPath(source_path),
                           options, dbg);
  if (!s.ok()) {
    return IOStatus::Corruption(
        "RollbackAuxDirRenameNoLock: Failed to roll back auxiliary path "
        "renaming");
  }

  return s;
}

/* Must hold files_mtx_ */
IOStatus ZenFS::RenameAuxPathNoLock(const std::string& source_path,
                                    const std::string& dest_path,
                                    const IOOptions& options,
                                    IODebugContext* dbg) {
  IOStatus s;
  std::vector<std::string> children;
  std::vector<std::string> renamed_children;

  s = target()->RenameFile(ToAuxPath(source_path), ToAuxPath(dest_path),
                           options, dbg);
  if (!s.ok()) {
    return s;
  }

  GetZenFSChildrenNoLock(source_path, true, &children);

  for (const auto& child : children) {
    s = RenameChildren(source_path, dest_path, child, options, dbg);
    if (!s.ok()) {
      IOStatus failed_rename = s;
      s = RollbackAuxDirRenameNoLock(source_path, dest_path, renamed_children,
                                     options, dbg);
      if (!s.ok()) {
        return s;
      }
      return failed_rename;
    }
    renamed_children.push_back(child);
  }

  return s;
}
```

Reply:
Sure that makes sense!